### ISSUE TITLE
Session being same for different clients.

### ISSUE THREAD

**Issue by shadow-fox** [expressjs/session#143]
Hi, I am facing this issue, (Which I have posted in Stackoverflow as well without any luck.). Hence reaching out here.

I have multiple node server serving requests. For each request without a session, I am creating a new session. Sometimes, an existing session id is being assigned to a new request. Before creating a unique session, I am checking whether the cookie is there...if it is, then I am not creating a new session. But while doing so, we are seeing that same session id being shared with different client.

Here is the function i.e. being called in `app.js` in `express.js` application.

```
var express         = require('express');
var session         = require('express-session');
var RedisStore      = require('connect-redis')(session);
var cookieParser    = require('cookie-parser');
var bodyParser      = require('body-parser');
var passport        = require('passport');

var app = express();
app.use(bodyParser.json());// to support JSON-encoded bodies
app.use(bodyParser.urlencoded({ extended: false }));
app.use(compress());
app.use(cookieParser());
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false }));

//app.use(cookieParser());
var redis = new RedisStore({
  host:config.redis.url,
  port:config.redis.port, 
  prefix:'sess-'+new Date().getDate()+'-'+(new Date().getMonth()+1)+'-'+new Date().getFullYear()+':'
});
app.use(genSess());
app.use(express.static(path.join(__dirname, 'public')));
app.use(passport.initialize());
app.use(passport.session());

function genSess() {
    return function (req, res, next) {
        if(/ucompany=/.test(req.headers['cookie'])){
            var cookie = req.headers['cookie'].split("ucompany=");
            var genid = cookie[1];
            return session({
                genid:function () {
                    return genid;
                },
                store: redis,
                cookie: {
                    maxAge: new Date(Date.now() + (7 * 24 * 60 * 60 * 1000))
                },
                secret: 'ucomp123',
                resave: false,
                name: "ucompany",
                saveUninitialized: true
            })(req, res, next)

        }
        return session({
            store: redis,
            cookie: {
                maxAge: new Date(Date.now() + (7 * 24 * 60 * 60 * 1000))
            },
            secret: 'ucomp123',
            resave: false,
            name: "ucompany",
            saveUninitialized: true
        })(req, res, next)
    }
}
```

**Comment by JoeWagner** [comment]
The proper use of this module is `app.use(session({/* options */}))`
What you are doing above creates a new session middleware for each request.  You only need one middleware, which should handle each request.  Also the `genid` option should generate a unique id for new sessions.  NOT return an existing id. [docs](https://github.com/expressjs/session#genid)

Try getting rid of your `genSess` function and just passing in one set of options to `express-session`, i.e.
```
app.use(session({
    store: redis,
    cookie: {
        maxAge: new Date(Date.now() + (7 * 24 * 60 * 60 * 1000))
    },
    secret: 'ucomp123',
    resave: false,
    name: "ucompany",
    saveUninitialized: true
}));
```

**Comment by bhupi-zivame** [comment]
Is this:

var redis = new RedisStore({
  host:config.redis.url,
  port:config.redis.port, 
  prefix:'sess-'+new Date().getDate()+'-'+(new Date().getMonth()+1)+'-'+new Date().getFullYear()+':'
});

the correct way to generate prefix of session id's?

**Comment by dougwilson** [comment]
@bhupi-zivame for that question, you'll want to point in the `connect-redis` repo.

**Comment by shadow-fox** [comment]
```
var express         = require('express');
var session         = require('express-session');
var RedisStore      = require('connect-redis')(session);
var cookieParser    = require('cookie-parser');
var bodyParser      = require('body-parser');
var passport        = require('passport');

var app = express();
app.use(bodyParser.json());// to support JSON-encoded bodies
app.use(bodyParser.urlencoded({ extended: false }));
app.use(compress());
app.use(cookieParser());
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false }));

//app.use(cookieParser());
var redis = new RedisStore({
  host:config.redis.url,
  port:config.redis.port, 
  prefix:'sess-'+new Date().getDate()+'-'+(new Date().getMonth()+1)+'-'+new Date().getFullYear()+':'
});
app.use(express.static(path.join(__dirname, 'public')));
app.use(session({
    store: redis,
    cookie: {
        expires: new Date(Date.now() + (7 * 24 * 60 * 60 * 1000)),
        maxAge:7 * 24 * 60 * 60 * 1000
    },
    secret: 'ucomp123',
    resave: false,
    name: "ucomapny",
    saveUninitialized: true
}));

app.use(passport.initialize());
app.use(passport.session());
```

I have changed my code to as per docs and what @JoeWagner said, but I am still seeing the same issue. Session id is not being unique, it has been creating same session id for two different client. (two completely different user, with different ip as well as different machine(browser) ).

And one thing I have noticed is that, while doing as per above, I am seeing sessions being created for bots as well which is huge in number, e.g. google,alexa etc are hitting on the server to crawl through. For that reason, I was creating  new session middleware.

**Comment by JoeWagner** [comment]
@shadow-fox The code above all looks correct.  I'll run this and see if I can reproduce the issue.  As for the bots, https://github.com/expressjs/session/issues/133 has two possible solutions

**Comment by JoeWagner** [comment]
@shadow-fox I tried reproducing your issue with no success.  Can you provide anymore info on the specifics of your setup, i.e. what you've tried and exactly what you are seeing?

**Comment by dougwilson** [comment]
Hi @shadow-fox any additional information to help us? We have not been successful in reproducing the issue and are unable to proceed without more information from you.

**Comment by shadow-fox** [comment]
Hi @dougwilson , this is what i have done to overcome the issue, (don't know whether it will help or not)
1. Added node-uuid to create guid. (we are using v4)
2. removed dynamic session key creation i.e. prefix for redis.
3. for bots i have done as per #133 describes.

```
var express         = require('express');
var session         = require('express-session');
var RedisStore      = require('connect-redis')(session);
var cookieParser    = require('cookie-parser');
var bodyParser      = require('body-parser');
var passport        = require('passport');
var uuid = require('node-uuid');

var app = express();
app.use(bodyParser.json());// to support JSON-encoded bodies
app.use(bodyParser.urlencoded({ extended: false }));
app.use(compress());
app.use(cookieParser());
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false }));

var redis = new RedisStore({
  host:config.redis.url,
  port:config.redis.port
});
app.use(express.static(path.join(__dirname, 'public')));
// For bots as mentioned in #133 issue
app.use(function(req, res, next){
  if (!req.headers['user-agent'] || req.headers['user-agent'] == "" || /^alexa|^blitz\.io|bot|^browsermob|crawl|^curl|^facebookexternalhit|feed|google web preview|^ia_archiver|^java|jakarta|^load impact|^magespeedtest|monitor|nagios|^pinterest|postrank|slurp|spider|uptime|yandex/.test(req.headers['user-agent'])){
            return next();
        }
});
app.use(session({
    genid:function () {
        return uuid.v4();
    },
    store: redis,
    cookie: {
        expires: new Date(Date.now() + (60 * 60 * 1000)),
        maxAge: 60 * 60 * 1000,
        secure: false
    },
    secret: 'ucomp123',
    resave: false,
    name: "ucomapny",
    saveUninitialized: true
}));

app.use(passport.initialize());
app.use(passport.session());
```

This solved the same session id being generated issue, not sure about the session being created for bots, is that correct ?

**Comment by shadow-fox** [comment]
@dougwilson  @JoeWagner :
 The set up of the servers:
1. Two node instance is being running the app on ec2. 
2. No nginx/apache/HAProxy.
3. Code gets deployed to both those servers.
4. Starts by forever.

The issue:
For some users I am seeing session being same. Tested and found that, when one user gets log in, he/she gets another user's details in his/her profile page. I store the email and username inside the session. If both session are different i wouldnt have had same profile info. So checked in the redis what was the issue, and saw that it is someone else session id. It just assigns the session id to this user.

**Comment by dougwilson** [comment]
In the end, we really cannot help much unless you remove the use of `passport` module from the equation, because it's very complicated, I don't know how it works, and has caused this exact issue for some other users in the past.

**Comment by shadow-fox** [comment]
@dougwilson I will try to do that. In the meanwhile as the original issue is not present in the system I am closing this issue. If in future occurs again I will create a new issue. For the part as you said about passport, I will try to do that in one of the server to check it out, but now I can say after using ```node-uuid``` I am not seeing this issue.

**Comment by larafale** [comment]
Hello, did you find out the real deal ?
we are having similar issue, we use heroku with 2 dynos running 
we use the following lib:
```
"connect-mongo": "0.8.1",
"express": "~4.10.2",
"express-session": "1.12.1"
```

we have noticed that with significant load on our site, some users managed to have another user session assigned, from a different IP. we save sessions only when initialized. 

this is our app
```
var session = require('express-session')
  , store = require('connect-mongo')({ session: session })
  , uuid = require('node-uuid')

app.use(session({ 
      secret: 'secret'
    , genid: function(){ return uuid.v4() } 
    , key: 'app.sid' 
    , cookie: { maxAge: 1000 * 60 * 60 * 24 * 5 } // 5days
    , store: new store({ url: config.store })
    , resave: true
    , saveUninitialized: false
  }))
```

and this is the only place in our app where we set session (only after a login, we never update afterward)
```
router
  .post('/session', function(req, res){
    req.session.ip = req.userIP
    req.session.token = req.body.token || req.session.token
    req.session.user = _.pick(req.body.user, '_id', 'name', 'nick', 'pic', 'fingerprint')
    res.status(200).end()
})
```

Somehow few requests sends back someone else session, this is bad shit :)
we can't figure out where the problem is taking place, express-session ? connect-mongo ? heroku server instance ? mongodb (we use verison 2.6.4 replicaset on compose.io) ?

well if any of you have any idea that would be great :) 
Thanks guys, keep up the good work, that's art we'are all doing !

**Comment by larafale** [comment]
ok we've found that f***** bug, we had a middleware assigning `app.locals.sessions` to `req.session` 
```
app.use(function(req, res, next){
    app.locals.sessions = req.session
    next()
  })
```
to pass session data to jade template like so : 

```
script.
  var $token = '!{session.token}' || false
```

`app.locals` is an express global variable, there is still a time between when `app.locals` is being assigned and the page is being rendered, and with concurrent requests, it can lead to situations where  `app.locals` have changed to another user session value when it is passed in the rendering process of the jade template.

There are many ways to avoid this problem, like using `res.locals` instead of `app.locals` and pass session data in the render function of your controller. But we wanted the session to be added in every render so we added this middleware which basically overided the render function passing the sessions directly in the render process : 
```
app.use(function(req, res, next){
    var render = res.render
      , sessions = req.session

    res.render = function(view, locals, cb){
      locals = locals || {}
      locals.session = sessions
      render.call(res, view, locals, cb)
    }
    next()
  })
```

Voilà :) we will try it tomorow but are pretty sure that's it !
hope this will help
